collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright 2018 Nafundi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.odk.collect.android.utilities;
+
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+
+import org.javarosa.xform.parse.XFormParser;
+import org.kxml2.kdom.Element;
+import org.odk.collect.android.R;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.dao.FormsDao;
+import org.odk.collect.android.logic.FormDetails;
+import org.odk.collect.android.logic.ManifestFile;
+import org.odk.collect.android.logic.MediaFile;
+import org.odk.collect.android.preferences.PreferenceKeys;
+import org.opendatakit.httpclientandroidlib.client.HttpClient;
+import org.opendatakit.httpclientandroidlib.protocol.HttpContext;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import timber.log.Timber;
+
+public class DownloadFormListUtils {
+
+    // used to store error message if one occurs
+    public static final String DL_ERROR_MSG = "dlerrormessage";
+    public static final String DL_AUTH_REQUIRED = "dlauthrequired";
+
+    private static final String NAMESPACE_OPENROSA_ORG_XFORMS_XFORMS_LIST =
+            "http://openrosa.org/xforms/xformsList";
+
+    private static boolean isXformsListNamespacedElement(Element e) {
+        return e.getNamespace().equalsIgnoreCase(NAMESPACE_OPENROSA_ORG_XFORMS_XFORMS_LIST);
+    }
+
+    private DownloadFormListUtils() {
+    }
+
+    public static HashMap<String, FormDetails> downloadFormList(boolean alwaysCheckMediaFiles) {
```

Comment:
I realise that this code has been moved from DownloadFormListTask, but I wonder if this is an opportunity to reduce the size and complexity of this method. It performs two different parsing methods all bundled up into one long method. A simple thing could be to create separate methods to handle the two cases. A more OO way might be to have a parsing interface and create the most appropriate parsing concrete implementation.

Reply:
Yes it's not my code and it should be fixed in a separate pr since this one is pretty big itself.